### PR TITLE
Support having secondary visibility store through `secondaryVisibilityStore` config key

### PR DESCRIPTION
**Pull request by hazelkim052**
## What changed?

- modified persistence validation to support having secondary visibility store through `secondaryVisibilityStore` config key
- user can set dual ES visibility via 
   1. the existing indices.secondary_visibility key (backwards compatibility); or
   2. using the secondaryVisibilityStore and setting another data store config
- restrictions:
    - they must be mutually exclusive
    - ES config must be same except for `indices.visibility` in case of using secondaryVisibilityStore config (this is because we still have the global ES client)

### CODE REVIEW

File: common/config/persistence.go [hunk 1]
```diff
@@ -41,6 +41,11 @@ const (
 	StoreTypeNoSQL = "nosql"
 )
 
+var (
+	ErrInvalidVisibilityStoreCombination = errors.New("persistence config: Cannot set visibilityStore and secondaryVisibilityStore with different datastore types")
+	ErrConfigMismatch                    = errors.New("persistence config: Config mismatch for visibilityStore and secondaryVisibilityStore")
+)
```
Comment (rodrigozhou) [inline common/config/persistence.go:47]: Create a base error, and then use below like this:
```
var ErrPersistenceConfig = errors.New("persistence config error")


return fmt.Errorf("%w: config mismatch for visibilityStore and secondaryVisibilityStore", ErrPersistenceConfig)
```

File: common/config/persistence.go [hunk 2]
```diff
@@ -68,18 +73,25 @@ func (c *Persistence) Validate() error {
...
 
 	if c.VisibilityStore == "" {
 		return errors.New("persistence config: visibilityStore must be specified")
 	}
-	if c.DataStores[c.VisibilityStore].Elasticsearch != nil && c.SecondaryVisibilityStore != "" {
+	if c.SecondaryVisibilityStore != "" {
+		if (c.DataStores[c.VisibilityStore].Elasticsearch == nil && c.DataStores[c.SecondaryVisibilityStore].Elasticsearch != nil) ||
+			(c.DataStores[c.VisibilityStore].Elasticsearch != nil && c.DataStores[c.SecondaryVisibilityStore].Elasticsearch == nil) {
+			return ErrInvalidVisibilityStoreCombination
+		}
```
Comment (rodrigozhou) [inline common/config/persistence.go:89]: Let's make this more readable:
```suggestion
		isPrimaryEs := c.DataStores[c.VisibilityStore].Elasticsearch != nil
		isSecondaryEs := c.DataStores[c.SecondaryVisibilityStore].Elasticsearch != nil
		if (isPrimaryEs != isSecondaryEs) {
			return ErrInvalidVisibilityStoreCombination
		}
```

File: common/config/persistence.go [hunk 3]
```diff
@@ -68,18 +73,25 @@ func (c *Persistence) Validate() error {
...
+	}
+	if c.DataStores[c.VisibilityStore].Elasticsearch.GetSecondaryVisibilityIndex() != "" && c.SecondaryVisibilityStore != "" {
```
Comment (rodrigozhou) [inline common/config/persistence.go:91]: You can add this to the previous condition which already checked if `c.SecondaryVisibilityStore != ""`.

File: common/config/persistence.go [hunk 4]
```diff
@@ -68,18 +73,25 @@ func (c *Persistence) Validate() error {
...
 		return errors.New(
 			"persistence config: cannot set secondaryVisibilityStore " +
-				"when visibilityStore is setting elasticsearch datastore",
+				"when visibilityStore is setting secondary visibility index",
```
Comment (rodrigozhou) [inline common/config/persistence.go:94]:
```suggestion
				"when visibilityStore is setting Elasticsearch secondary visibility index",
```

File: common/config/persistence.go [hunk 5]
```diff
@@ -99,19 +109,21 @@ func (c *Persistence) Validate() error {
...
+		visibilityStoreEsConfig := reflect.ValueOf(c.DataStores[c.VisibilityStore].Elasticsearch).Elem()
+		secondaryVisibilityStoreEsConfig := reflect.ValueOf(c.DataStores[c.SecondaryVisibilityStore].Elasticsearch).Elem()
+		for i := 0; i < max(visibilityStoreEsConfig.NumField(), secondaryVisibilityStoreEsConfig.NumField()); i++ {
+			if visibilityStoreEsConfig.Type().Field(i).Name == "Indices" || visibilityStoreEsConfig.Type().Field(i).PkgPath != "" {
+				continue
+			}
+			if !reflect.DeepEqual(visibilityStoreEsConfig.Field(i).Interface(), secondaryVisibilityStoreEsConfig.Field(i).Interface()) {
+				return fmt.Errorf("%w: %q", ErrConfigMismatch, visibilityStoreEsConfig.Type().Field(i).Name)
+			}
+		}
```
Comment (rodrigozhou) [inline common/config/persistence.go:125]: This looks unnecessarily complex. Wouldn't this work?

```suggestion
		esConfig := *c.DataStores[c.VisibilityStore].Elasticsearch
		secEsConfig := *c.DataStores[c.SecondaryVisibilityStore].Elasticsearch
		esConfig.Indices = nil
		secEsConfig.Indices = nil
		if !reflect.DeepEqual(esConfig, secEsConfig) {
			return ErrConfigMismatch
		}
```

**Review by rodrigozhou — CHANGES_REQUESTED** [review]

File: common/config/persistence.go (hunk 1 above)
Comment (hazelkim052) [reply common/config/persistence.go:47]: I created a base error to be "persistence config" as we are wrapping the validation error already. This is what error looks like: `config validation error: persistence config: error messages`

File: common/config/persistence.go (hunk 5 above)
Comment (hazelkim052) [reply common/config/persistence.go]: yes this should also work, I updated and tested the code. Thanks for the suggestion!

File: common/config/persistence.go [hunk 6]
```diff
@@ -68,50 +70,60 @@ func (c *Persistence) Validate() error {
...
+				ErrPersistenceConfig)
+		}
+		if c.DataStores[c.SecondaryVisibilityStore].Elasticsearch.GetSecondaryVisibilityIndex() != "" {
+			return fmt.Errorf(
+				"%w: secondary visibility datastore %q cannot set secondary_visibility",
+				ErrPersistenceConfig,
+				c.SecondaryVisibilityStore)
+		}
+		// ElasticSearch config for visibilityStore and secondaryVisibilityStore must be the same except for
+		// `indices.visibility` config key and private fields - this is a restriction due to global ES client
```
Comment (rodrigozhou) [inline common/config/persistence.go:103]: Nit: Move the comments to inside the if block.

**Review by rodrigozhou — APPROVED** [review]

**Pull request merged**